ARCs/arc-0004.md

```diff
@@ -0,0 +1,498 @@
+---
+arc: 4
+title: Algorand Application Binary Interface (ABI)
+status: Draft
+---
+
+# Algorand Transaction Calling Conventions
+
+> This API is a draft.  Some elements may change.
+
+## Summary
+
+A standard for encoding contract call transactions to invoke methods
+on an Algorand _app_ (stateful contract).
+
+## Abstract
+
+This document introduces conventions for encoding method calls,
+including argument and return value encoding, in Algorand application
+call transactions. The goal is to allow clients, such as wallets and
+dapps, to properly encode call transactions based on a simple
+description of the interface. Further, explorers will be able to show
+details of these method invocations.
+
+## Definitions
+
+* **HLL:** a higher level language that compiles to TEAL bytecode.
+* **dapp**: a decentralized application, interpreted here to mean an
+  off-chain application (a webapp, native app, etc.) that interacts
+  with smart contracts on the blockchain.
+* **wallet**: an application that stores secret keys for on-chain
+  accounts and can display and sign transactions for these accounts.
+* **explorer**: an application that allows browsing the blockchain,
+  showing details of transactions.
+
+## Specification
+
+The key words "**MUST**", "**MUST NOT**", "**REQUIRED**", "**SHALL**",
+"**SHALL NOT**", "**SHOULD**", "**SHOULD NOT**", "**RECOMMENDED**",
+"**MAY**", and "**OPTIONAL**" in this document are to be interpreted
+as described in [RFC-2119](https://www.ietf.org/rfc/rfc2119.txt).
+
+> Comments like this are non-normative.
+
+### Overview
+
+This document makes recommendations for encoding method invocations as
+application call transactions, and for describing methods for
+access by higher-level entities.  Encoding recommendations are
+intended to be minimal, intended only to allow interoperability among
+contracts.  Higher level recommendations are intended to enhance
+user-facing interfaces, such as high-level language, dapps, and
+wallets.  Apps that follow the recommendations described here are called
+_ARC-4 Apps_.
+
+### Methods
+
+A method is a section of code intended to be invoked externally with
+an application call transaction. A method must have a name, it may
+take a list of arguments as input when it is invoked, and it may
+return a single value (which may be a tuple) when it finishes
+running. The possible types for arguments and return values are
+described later.
+
+Invoking a method involves creating an application call transaction to
+specifically call that method. Methods are different from internal
+"helper functions" that may exist in a contract, but are not
+externally callable. In the future, applications may be able to invoke
+methods from other applications.
+
+
+#### Method Signature
+
+A method signature is a unique identifier for a method. The signature
+is a string that consists of the method's name, an open parenthesis, a
+comma-separated list of the types of its arguments, a closing
```

Comment:
Fair enough, @fabrice102 has been getting on me about that for a while.

Reply:
Are we allowing optional arguments? If so how do we handle them? Or maybe insisting that all arguments be there, but some may have a null value?

Reply:
No optional arguments as of yet.  Does anyone want to speak up in favor of them?  I'd prefer to avoid the complexity of specification until someone thinks they are important.